// package-control/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _axios = require('axios');

var _axios2 = _interopRequireDefault(_axios);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

async function fetchPackages(url) {
  const baseURL = 'https://packagecontrol.io/';
  const response = await (0, _axios2.default)({ baseURL, url, responseType: 'json' });
  const packages = response.data.packages;
  return packages;
}

class PackageControl {
  async search(query, opts = {}) {
    const { labels, limit } = opts;
    const packages = await fetchPackages(`search/${query}.json`);

    if (!labels) {
      return packages.slice(0, limit);
    }

    let searchLabels;

    if (typeof labels === 'string') {
      searchLabels = [labels];
    }

    return packages.filter(pkg => searchLabels.every(label => pkg.labels.includes(label))).slice(0, limit);
  }

  async getColorSchemes(limit) {
    const packages = await fetchPackages('browse/labels/color scheme.json');
    return packages.slice(0, limit);
  }

  async getThemes(limit) {
    const packages = await fetchPackages('browse/labels/theme.json');
    return packages.slice(0, limit);
  }

  async getLanguages(limit) {
    const packages = await fetchPackages('browse/labels/language syntax.json');
    return packages.slice(0, limit);
  }
}

exports.default = PackageControl;
